cumulusci/tasks/bulkdata/extract.py

```diff
@@ -91,90 +89,81 @@ def _init_mapping(self):
-            if sf == "Records not found for this query":
-                return
-            if sf:
-                column = mapping.get("fields", {}).get(sf)
-                if not column:
-                    lookup = mapping.get("lookups", {}).get(sf, {})
-                    if lookup:
-                        lookup_keys.append(sf)
-                        column = get_lookup_key_field(lookup, sf)
-                if column:
-                    columns.append(column)
+        for sf in fields:
+            column = mapping.get("fields", {}).get(sf)
+            if not column:
+                lookup = mapping.get("lookups", {}).get(sf, {})
+                if lookup:
+                    lookup_keys.append(sf)
+                    column = get_lookup_key_field(lookup, sf)
+            if column:
+                columns.append(column)
+
         if not columns:
             return
         record_type = mapping.get("record_type")
         if record_type:
             columns.append("record_type")
 
-        processor = log_progress(
-            process_incoming_rows(result_file, record_type), self.logger
-        )
-        data_file = IteratorBytesIO(processor)
+        record_iterator = step.get_results()
+        if record_type:
+            record_iterator = map(lambda rec: rec + [record_type], record_iterator)
+
         if mapping["oid_as_pk"]:
-            self._sql_bulk_insert_from_csv(conn, mapping["table"], columns, data_file)
+            self._sql_bulk_insert_from_records(
+                conn, mapping["table"], columns, record_iterator
+            )
         else:
-            # If using the autogenerated id field, split out the CSV file from the Bulk API
+            # If using the autogenerated id field, split out the returned records
             # into two separate files and load into the main table and the sf_id_table
-            with tempfile.TemporaryFile("w+b") as f_values:
-                with tempfile.TemporaryFile("w+b") as f_ids:
+
+            with tempfile.TemporaryFile("w+") as f_values:
+                with tempfile.TemporaryFile("w+") as f_ids:
                     data_file_values, data_file_ids = self._split_batch_csv(
-                        data_file, f_values, f_ids
+                        record_iterator, f_values, f_ids
                     )
-                    self._sql_bulk_insert_from_csv(
-                        conn, mapping["table"], columns, data_file_values
+                    self._sql_bulk_insert_from_records(
+                        conn, mapping["table"], columns, csv.reader(data_file_values)
                     )
-                    self._sql_bulk_insert_from_csv(
-                        conn, mapping["sf_id_table"], ["sf_id"], data_file_ids
+                    self._sql_bulk_insert_from_records(
+                        conn,
+                        mapping["sf_id_table"],
+                        ["sf_id"],
+                        csv.reader(data_file_ids),
```

Comment:
I have not benchmarked the change.

More generally, I'm very eager to decouple as much as possible from the details of underlying database storage. The complexity of the logic around database interaction makes it hard to significantly extend the Bulk Data toolchain in the directions we want to go.

In specific, I asked @prescod if he'd ever used that functionality and he said no, but that's perhaps not enough to justify its removal. More broadly, with Paul's work on Snowfakery in place, do we have a use case for storing a static LDV dataset in Postgres? I believe only NPSP's original LDV data set is stored that way at present, and I don't believe the Postgres database has been used since I started at SFDO.

Reply:
I agree that decoupling logic from the underlying data storage is a good goal -- I just want to make sure we're considering the implications. It sounds like removing the special postgres handling won't affect how we are doing things these days, but let me fill in a bit more historic context.

In late 2018 when I started to use Jason's original bulkdata tasks to load LDV test data for testing DSO, performance was an issue. This was true for both phases of the process: generating data in the local db, and loading it from there into Salesforce. When I profiled the latter in particular, I found that uploading the batches was quick, processing the batches was out of our control (aside from switching to parallel batches) and downloading the results was quick, but loading the result ids into the local db was slow. One change I made as a result was to put the ids in a separate table -- this allowed us to do an insert instead of update, which helped a lot. I also did quite a bit of experimentation to figure out what the fastest way was to get the ids inserted, and that's what led to using postgres rather than sqlite and using postgres' COPY FROM feature rather than sqlalchemy to load the ids.

What I'm trying to say is that I'll sign off on this refactoring even if we're losing the special postgres support, but:
1. I want to make sure Paul's aware that there's probably better performance available for loading result ids with an approach like this.
2. I would like to see a benchmark comparison for loading a significant-sized dataset before and after these changes, to make sure we're not introducing significant regressions that are unrelated to the special postgres support.

Reply:
I do have a long-range goal of storing some large data sets in cloud-hosted Postgres some day: partially because of performance and partially because of repeatability. I don't have first-hand experience with the performance implications of the optimizations and its hard for me to have a strong opinion until one of us has a chance to do before and after tests. My most recent experiences would tend to colour my opinion towards "Salesforce is 99% of the problem and SQL is 1%" but in a situation with more reasonable data skew or some other workaround, that might not be true anymore. Salesforce is multi-threaded and this code is single-threaded, so Salesforce import might actually be more reasonable if I could use it as it was designed to be used.

I can imagine an abstraction that would allow the caller to detect that the recipient prefers CSV and it would serve CSV (not unlike content-negotiation on the Web). You'd need a Postgres sub-class of the SQL Alchemy mixin, I guess (or some smarts in the mixin analogous to the old code).

Maybe it's not much more crazy than

if self._storage_prefersCSV():
   self._import_from_CSV()
else:
   self._import_from_records()

But on the other hand, I don't totally understand what (present or future) refactorings are made difficult by having CSV as the standardized interchange format so perhaps I'm blocking the same refactorings by making it an option.

But if David Reed sees it as too messy then I can treat it as a "future optimization opportunity" and deal with it when the time comes.

Reply:
Memorializing our conversation this morning: I'm going to add the original Postgres optimization back in (dumping the record generator to a CSV file first), and we'll see if we can benchmark that.

Reply:
I've re-added the Postgres optimization and added some unit tests. @prescod Do you have any active Postgres data sets now that could be used to test it end-to-end?

Reply:
@davisagli : I think you have a large Postgres dataset around somewhere?

Reply:
@prescod alas, not any more

Reply:
Hmm. If we don't have a way to test it, what about segregating this optimization off into a branch for the time down the line when @prescod starts using Postgres infrastructure? I don't want to ship a code path we can't test, and doing a lot of work to set up Postgres to test an optimization we are not using feels misplaced.